File: android/app/src/main/java/com/dtools/AppWidget/ImageAppWidgetProvider.java
package com.dtools.AppWidget;

import android.appwidget.*;
import android.content.*;
import android.graphics.*;
import android.widget.*;
import com.dtools.R;
import com.dtools.Storage.*;

public class ImageAppWidgetProvider extends AppWidgetProvider {
    @Override
    public void onDeleted(Context context, int[] appWidgetIds) {
        super.onDeleted(context, appWidgetIds);
        for (int id:appWidgetIds) {
            AndroidSharedPreferences.set(context, String.valueOf(id), "");
        }
    }

    @Override
    public void onDisabled(Context context) {
        super.onDisabled(context);
    }

    @Override
    public void onEnabled(Context context) {
        super.onEnabled(context);
    }

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        super.onUpdate(context, appWidgetManager, appWidgetIds);
        Bitmap defaultValue = BitmapFactory.decodeResource(context.getResources(), R.drawable.image_app_widget_preview);
        for (int id:appWidgetIds) {
            Bitmap bitmap = AndroidSharedPreferences.getBitmap(context, String.valueOf(id), defaultValue);
            RemoteViews remoteViews = new RemoteViews(context.getPackageName(), R.layout.image_app_widget);
            remoteViews.setImageViewBitmap(R.id.image_app_widget, bitmap);
            AppWidgetManager.getInstance(context).updateAppWidget(id, remoteViews);
        }
    }
    @Override
    public void onReceive(Context context, Intent intent) {
        super.onReceive(context, intent);
    }
}

File: android/app/src/main/kotlin/com/dtools/AppWidget/ImageAppWidgetProvider.kt
package com.dtools.AppWidget

import android.appwidget.*
import android.content.*
import android.graphics.*
import android.widget.*
import com.dtools.R
import com.dtools.Storage.*

class ImageAppWidgetProvider : AppWidgetProvider() {
    override fun onDeleted(context: Context, appWidgetIds: IntArray) {
        super.onDeleted(context, appWidgetIds)
        for (id in appWidgetIds) {
            AndroidSharedPreferences.set(context, id.toString(), "")
        }
    }

    override fun onDisabled(context: Context) {
        super.onDisabled(context)
    }

    override fun onEnabled(context: Context) {
        super.onEnabled(context)
    }

    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray) {
        super.onUpdate(context, appWidgetManager, appWidgetIds)
        val defaultValue = BitmapFactory.decodeResource(context.resources, R.drawable.image_app_widget_preview)
        for (id in appWidgetIds) {
            val bitmap = AndroidSharedPreferences.getBitmap(context, id.toString(), defaultValue)
            val remoteViews = RemoteViews(context.packageName, R.layout.image_app_widget)
            remoteViews.setImageViewBitmap(R.id.image_app_widget, bitmap)
            AppWidgetManager.getInstance(context).updateAppWidget(id, remoteViews)
        }
    }

    override fun onReceive(context: Context, intent: Intent) {
        super.onReceive(context, intent)
    }
}
